mininet/node.py

```diff
@@ -839,6 +839,42 @@ def init( cls ):
         cls.inited = True
 
 
+class XIAHost( Host ):
+    "eXpressive Internet Architecture host"
+
+    def __init__( self, name, **kwargs ):
+        Node.__init__( self, name, **kwargs )
+        self.xids = {}  # mapping of configurable inputs to XID types
+
+    def config( self, hid=None, **params):
+        """hid: private key file names for Host Identifiers eg. hid='pk1,pk2'
+           params: parameters for Node.config()"""
+        r = Node.config( self, **params )
+        self.setParam( r, 'setHID', hid=hid )
```

Comment:
In `examples/echoxp.py`, we call `addHost()` which puts the `hid` parameter inside of `params`. Don't we need to get the `hid` parameter out of `params` to pass as the argument to `setHID()` here? Otherwise, `setHID()` will always be called with `hid=None`, since that's the default value for `hid` inside of this function `config()`.

In other words, shouldn't it be:

    self.setParam( r, 'setHID', hid=params['hid'] )

Then we can drop the `hid=None` parameter to this config function. This will be especially good for when we add more principal types, so that we don't need to continue adding parameters to this function. It could look like:

    def config( self, **params):
         """hid: private key file names for Host Identifiers eg. hid=['foo', 'bar']
            params: parameters for Node.config()"""
        r = Node.config( self, **params )
        if 'hid' in params:
            self.setParam( r, 'setHID', hid=params['hid'])
        if 'xdp' in params:
            self.setParam( r, 'setXDP', xdp=params['xdp'])

Reply:
Suppose we have `addHost(xia1, cls=XIAHost, hid='xia1_hid1')` 
In  `def config( self, hid=None, **params):`  hid is separated out of the dictionary and the residue keyworded arguments are stored in params. Hence, the default value of hid is overwritten as xia1_hid1.

(it's slightly different this keyworded variable length parameters of python than ordered argument passing)

This residue is used to do the default configurations by calling Node.config(**params)
this **params no longer has the hid, because we have taken that out already

Reply:
If you specify a keyword argument `hid` in a call like `addHost(xia1, cls=XIAHost, hid='xia1_hid1')`, then when `def config( self, hid=None, **params)` is invoked, `hid` will be a keyword in the dictionary `params`.

What I'm saying is that we don't need to add an optional parameter `hid` to be able to tell whether or not an `hid` keyword was specified. We can know that it wasn't specified by saying something like `if 'hid' not in params`.

Reply:
The benefit of keeping the declaration just `def config( self, **params)` is that we don't need to add an optional argument for each principal type that we add configuration for.